2019-03-Autumn/0_arrays_sorting/README.md

```diff
@@ -86,20 +92,80 @@ TODO :bug:
 
 TODO :bug:
 
-## 3. PROBLEM 3 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution 
-
-TODO :bug:
-
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
-
-
+### 3. Sub Sort
+
+Source: CTCI 16.16
+
+#### Solution
+
+###### Steps
+
+To solve this problem, we must find a starting point for our middle subarray 
+which is unsorted. The easiest way to do this is to check the LHS and RHS 
+subarrays for the longest fully sorted segment. (For now, we will not think about 
+whether the LHS and RHS are sorted relative to each other.)
+
+Having found two index markers for our inner, unsorted array (anywhere from 
+two elements to the entire array minus two elements), we must determine what 
+the largest and smallest elements in our unsorted array are. These two values 
+will determine how far we have to expand the bounds into the LHS and RHS subarrays.
+
+Then, we expand outwards, comparing our `unsorted_min` to `arr[lhs]` and decrementing `lhs` until `arr[lhs] <= unsorted_min`. Likewise, `unsorted_max` must be less than `arr[rhs]`.
+
+###### Implementation
+
+```java
+public static int[] subsort(int[] array) {
+    // edge case
+    // if array is empty or length 1, no sorting is required. 
+    // return -1 as indeces (or null, if the client can guarantee null checking)
+    if (array.length <= 1) return new int[]{-1, -1};
+
+    // find bounds for sorted LHS 
+    int lhs = getSortedLhs(array);
+    // if LHS spans entire array, we can return a minimal window of 0 length
+    if (lhs >= array.length - 1) return new int[]{-1, -1};
+
+    // find bounds for sorted RHS
+    int rhs = getSortedRhs(array);
+    // find min and max in center section
+    int min = array[lhs];
+    int max = array[rhs];
+    for (int i = lhs; i <= rhs; i++) {
+        if (array[i] < min) min = array[i];
+        if (array[i] > max) max = array[i];
+    }
+    // Shrink LHS and RHS segments so that 
+    // all LHS items are less than min and 
+    // all RHS items are greater than max.
+    while (lhs > 0 && array[lhs - 1] > min) lhs--;
+    while (rhs < array.length - 2 && array[rhs + 1] < max) rhs++;
+    // Fencepost case for rhs index; if rhs gets incremented to 
+    // final array index, the truth condition above fails and the 
+    // value of the last index place never gets checked.
+    // If the desired bounds were not inclusive bounds, this would 
+    // not be necessary.
+    // (see [ 2, 1, 1 ] example)
+    if (rhs == array.length - 2 && array[rhs + 1] < max) rhs++;
```

Comment:
They can be condensed to:
while (rhs < array.length - 1 && array[rhs + 1] < max) rhs++;
If I remember correctly

Reply:
Lol, I don't know why my brain didn't realize this when the `lhs` portion makes total sense. Zero-based indexing wins again.